[wpilib] Change opmodes to purely periodic (#8652)
wpilibsuite/allwpilib: design-docs/opmodes.md

### OpModeRobot

> The `OpModeRobot` class is the base class for the user's `Robot` class. It extends `RobotBase` directly and implements the private library machinery for robot startup and robot execution (including creating and transitioning between opmodes in accordance with the opmode lifecycle, as described in the following section).

Reviewer: ```suggestion
The `OpModeRobot` class is the base class for the user's `Robot` class.  It extends `RobotBase` directly and implements the private library machinery for robot startup and robot execution (including creating and transitioning between opmodes in accordance with the opmode lifecycle, as described in the following section).
```
Double space between sentences is the convention for this design doc.

- While enabled, `periodic()` is called at `OpModeRobot#getPeriod()`, and additional callbacks from `getCallbacks()` are run at their own configured rates
- If robot disables or a different opmode is selected while enabled, `end()` is called then `close()` is called (Java), or the object is destroyed (C++/Python); the object is not reused
- If a different opmode is selected while disabled, only `close()` is called (Java), or the object is destroyed (C++); the object is not reused

Following `close()` being called (Java)/the opmode being destroyed (C++), a *new* opmode object is constructed based on the DS teleop/auto/test/match selector and selected opmode.  In teleop/auto/test, the drop-down selection will be the same as before the previous enable, so the same opmode class is constructed again.  In match (or when FMS-connected), only the selected auto opmode object is initially constructed; once auto completes, the selected teleop opmode object is constructed.  Thus only zero or one opmode objects will ever be "alive" at any given time.

For consistency in operation, the library will ensure that `disabledPeriodic()` is always called at least once before `start()` is called.

> User implementations of opmode classes may have either a no-parameter constructor or a constructor that accepts any subclass of `OpModeRobot`.  If available, the library will call the latter and pass the user's `Robot` object to it when constructing the class.

Reviewer: ```suggestion
User implementations of opmode classes may have either a no-parameter constructor or a constructor that accepts (a subclass of) the user's `Robot` class type.  If available, the library will call the latter and pass the user's `Robot` object to it when constructing the class.
```
The check is performed against the user's `Robot` class, not against `OpModeRobot`.

The `OpModeRobot` class is the base class for the user's `Robot` class.  It also implements the private library machinery for robot startup and robot execution (including creating and transitioning between opmodes in accordance with the opmode lifecycle, as described in the following section).

```java
public abstract class OpModeRobot extends TimedRobot {
  public void nonePeriodic() {
    // this code is called when no opmode is selected
  }

  // add opmodes with explicit registration
  public void addOpModeFactory(Supplier<OpMode> factory, RobotMode mode,
      String name, String group, String description,
      Color textColor, Color backgroundColor) {...}

>   public void addOpMode(Class<? extends OpMode> cls, RobotMode mode,
>       String name, String group, String description,
>       Color textColor, Color backgroundColor) {...}
>   public void addAnnotatedOpMode(Class<? extends OpMode> cls) {...}

Reviewer: ```suggestion

  // add a particular opmode class
  public void addOpMode(Class<? extends OpMode> cls, RobotMode mode,
      String name, String group, String description,
      Color textColor, Color backgroundColor) {...}
  public void addAnnotatedOpMode(Class<? extends OpMode> cls) {...}
```
It might be nice to move these into a different section since these two are Java-only.

  // add all annotated opmodes in a package and nested packages
  public void addAnnotatedOpModeClasses(Package pkg) {...}

  public void removeOpMode(RobotMode mode, String name) {...}
  public void publishOpModes() {...}
  public void clearOpModes() {...}
}
```

`OpModeRobot` (in Java) automatically scans the user's package and subpackages for `@Autonomous`, `@Teleop`, and `@TestOpMode` classes in its constructor and publishes that list to the DS.  When opmodes are added or removed manually after construction, `publishOpModes()` must be called to push changes to the DS.

### OpMode Classes

The `OpMode` interface serves as the base interface for all opmodes.  Most users will extend `PeriodicOpMode`, but users may also implement `OpMode` directly for custom behavior.

The lifecycle of an opmode is:

> - Operator selects opmode on DS → opmode object is constructed

Reviewer: ```suggestion
- When operator selects opmode on DS, a new opmode object is constructed
```
More consistent with the other items in the list. ("\<condition\>, \<event\>")

- While selected and disabled, `disabledPeriodic()` is called
- On disabled → enabled transition, `start()` is called once
- While enabled, `periodic()` is called at `OpModeRobot#getPeriod()`, and additional callbacks from `getCallbacks()` are run at their own configured rates

> - If robot disables or a different opmode is selected while enabled, `end()` is called asynchronously to request prompt termination, then `close()` is called; the object is not reused

Reviewer: Is `end()` going to be called asynchronously? I remember that there was discussion about it on Discord, but I don't remember the consensus. Note that right now it seems like the impl is synchronous (though of course that can change).

Reviewer 2: No, it shouldn't be.

- If a different opmode is selected while disabled, only `close()` is called; the object is not reused

Following `opModeClose()` being called, a *new* opmode object is constructed based on the DS teleop/auto/test/match selector and selected opmode.  In teleop/auto/test, the drop-down selection will be the same as before the previous enable, so the same opmode class is constructed again.  In match (or when FMS-connected), only the selected auto opmode object is initially constructed; once auto completes, the selected teleop opmode object is constructed.  Thus only zero or one opmode objects will ever be "alive" at any given time.

> For consistency in operation, the library will ensure that `disabledPeriodic()` is always called at least once before `start()` and is called.

Reviewer 2: ```suggestion
For consistency in operation, the library will ensure that `disabledPeriodic()` is always called at least once before `start()` is called.
```

- If a different opmode is selected while disabled, only `close()` is called; the object is not reused

Following `opModeClose()` being called, a *new* opmode object is constructed based on the DS teleop/auto/test/match selector and selected opmode.  In teleop/auto/test, the drop-down selection will be the same as before the previous enable, so the same opmode class is constructed again.  In match (or when FMS-connected), only the selected auto opmode object is initially constructed; once auto completes, the selected teleop opmode object is constructed.  Thus only zero or one opmode objects will ever be "alive" at any given time.

For consistency in operation, the library will ensure that `disabledPeriodic()` is always called at least once before `opModeRun()` is called.
For consistency in operation, the library will ensure that `disabledPeriodic()` is always called at least once before `start()` is called.

User implementations of opmode classes may have either a no-parameter constructor or a constructor that accepts the user's `Robot` class type.  If available, the library will call the latter and pass the user's `Robot` object to it when constructing the class.

> The library will use escalating steps to attempt to terminate a opmode if `opModeRun()` does not return within a reasonable timeframe after `opModeStop()` is called, up to and including termination of the robot executable process (which will result in an automatic restart of it at the system level).  User code (particularly in `LinearOpMode` derived classes) should check for `isRunning()` returning false and return as quickly as possible to allow the opmode to terminate gracefully.

Reviewer: It might be nice to keep this first sentence (With a note that only custom `OpMode` implementations will need to worry about it)- Knowing generally what happens if a custom `OpMode` implementation doesn't quit is pretty important, even if very few people need to worry about it.

Author: This is no longer necessary; once an opmode ends, its `periodic` method is simply not called again, as opposed to the robot trying to signal `opmodeRun` to exit.

Reviewer: Ah, that makes sense- I think I might've first made the comment before we moved the callback handling to `OpModeRobot` and then didn't notice that it was no longer relevant.
Good to be resolved.

> User implementations of opmode classes may have either a no-parameter constructor or a constructor that accepts any subclass of `OpModeRobot`.  If available, the library will call the latter and pass the user's `Robot` object to it when constructing the class.

Reviewer: ```suggestion
User implementations of opmode classes may have either a no-parameter constructor or a constructor that accepts (a subclass of) the `Robot` class.  If available, the library will call the latter and pass the user's `Robot` object to it when constructing the class.
```

> User implementations of opmode classes may have either a no-parameter constructor or a constructor that accepts any subclass of `OpModeRobot`.  If available, the library will call the latter and pass the user's `Robot` object to it when constructing the class.

Reviewer: > or a constructor that accepts any subclass of `OpModeRobot`.

This isn't quite right- The actual check is that there's a constructor which accepts specifically the instantiated `Robot`'s class.
https://github.com/wpilibsuite/allwpilib/blob/f19641829770f71f0603529b23462d496016cf49/wpilibj/src/main/java/org/wpilib/framework/OpModeRobot.java#L59-L63
https://github.com/wpilibsuite/allwpilib/blob/f19641829770f71f0603529b23462d496016cf49/wpilibj/src/main/java/org/wpilib/framework/OpModeRobot.java#L73-L75

Test opmode:

```java
@TestOpMode(name="Blink dashboard indicator")
public class TestDashboardIndicator extends PeriodicOpMode {
  private final Timer timer = new Timer();
  private boolean m_logged;

  @Override
  public void start() {

>     timer.restart();

Reviewer 2: restart or just start?

    m_logged = false;
  }

  @Override
  public void periodic() {
    if (!m_logged && timer.hasElapsed(0.5)) {
      Telemetry.log("indicator", true);
      Telemetry.log("indicator", false);
      m_logged = true;

>     }

Reviewer: ```suggestion
    if (!m_logged) {
      Telemetry.log("indicator", true);
      m_logged = true;
    }
    if (m_logged && timer.hasElapsed(0.5)) {
      Telemetry.log("indicator", false);
    }
```
Better matches the previous code. It does spam `false` at the end- To fix that, we'd need to distinguish between 3 separate states (beginning, waiting, and done), which could be done with another boolean or switching to an integer.
